homeassistant/components/evohome.py

```diff
@@ -0,0 +1,667 @@
+
+    except requests.RequestException as err:
+        if str(HTTP_BAD_REQUEST) in str(err):
+            # this happens when bad user credentials are supplied
+            _LOGGER.error(
+                "Failed to establish a connection with evohome web servers, "
+                "Check your username (%s), and password are correct."
+                "Unable to continue. Resolve any errors and restart HA.",
+                domain_data['params'][CONF_USERNAME]
+            )
+        else:
+            # Otherwise, it may be enough to back off and try again later.
+            raise PlatformNotReady(err)
+
+    finally:  # Redact username, password as no longer needed.
+        domain_data['params'][CONF_USERNAME] = 'REDACTED'
+        domain_data['params'][CONF_PASSWORD] = 'REDACTED'
+
+    domain_data['client'] = client
+
+    # Redact any installation data we'll never need.
+    if client.installation_info[0]['locationInfo']['locationId'] != 'REDACTED':
+        for loc in client.installation_info:
+            loc['locationInfo']['locationId'] = 'REDACTED'
+            loc['locationInfo']['streetAddress'] = 'REDACTED'
+            loc['locationInfo']['city'] = 'REDACTED'
+            loc['locationInfo']['locationOwner'] = 'REDACTED'
+            loc[GWS][0]['gatewayInfo'] = 'REDACTED'
+
+    # Pull down the installation configuration...
+    loc_idx = domain_data['params'][CONF_LOCATION_IDX]
+
+    try:
+        domain_data['config'] = client.installation_info[loc_idx]
+
+    # IndexError: configured location index is outside the range
+    except IndexError:
+        _LOGGER.warning(
+            "setup(): Config parameter, '%s'= %s , is out of range (0-%s), "
+            "continuing with '%s' = 0.",
+            CONF_LOCATION_IDX,
+            loc_idx,
+            len(client.installation_info) - 1,
+            CONF_LOCATION_IDX
+        )
+
+        domain_data['params'][CONF_LOCATION_IDX] = 0
+        domain_data['config'] = client.installation_info[0]
+
+    domain_data['status'] = {}
+
+    if _LOGGER.isEnabledFor(logging.DEBUG):
+        _LOGGER.debug(
+            "setup(): Location/TCS (temp. control system) used is: %s [%s]",
+            domain_data['config'][GWS][0][TCS][0]['systemId'],
+            domain_data['config']['locationInfo']['name']
+        )
+        # Some platform data needs further redaction before being logged.
+        tmp = dict(domain_data['config'])
+        tmp['locationInfo']['postcode'] = 'REDACTED'
+
+        _LOGGER.debug("setup(): domain_data['config']: %s", tmp)
+
+    # We have the platofrom configuration, but no state as yet, so...
+    def _first_update(event):
+        """Let the controller know it can obtain it's first update."""
+    # Send a message to the hub to do its first update()
+        pkt = {
+            'sender': 'setup()',
+            'signal': 'update',
+            'to': EVO_MASTER
+        }
+        hass.helpers.dispatcher.async_dispatcher_send(
```

Comment:
`hass.helpers.dispatcher.dispatcher_send`

Reply:
Why not async?  I thought async was 'a good thing'.  

I also use `async_schedule_update_ha_state` and `async_dispatcher_connect` - do I switch them over too?

Reply:
Async is a good thing, but you have to make sure you are in the right context for it. Methods and functions that are prefixed `async_` _must_ be run in the event loop. `_first_update` will be run in an executor thread so you need to use the synchronous helpers. Or decorate it `@callback` which will indicate to hass that it is safe to run in the event loop, then `async_` methods are okay. It's all about how things get scheduled. If you haven't seen it already, this might help: https://developers.home-assistant.io/docs/en/asyncio_categorizing_functions.html

Reply:
Too complicated for me - I will remove the three asyncs for now.

Reply:
Actually, two - my `_connect()` method has a `@callback`, OK? 
```python
    @callback
    def _connect(self, packet):
        """Process a dispatcher connect."""
        if packet['to'] & self._type and packet['signal'] == 'update':
            self.async_schedule_update_ha_state(force_refresh=True)
```

Reply:
seems OK to leave this in